Improve performance of repeated calls to popup.hide()

hughsk opened the pull request:
Previously, _resizeConstrainer() was being called every time popup.hide() was being called. This introduces a simple check that prevents popup.hide() from running any heavy code if it's already hidden.

This improves the performance of the popup module in the point map template, which currently calls popup.hide() once per frame. It's possible to make changes in the template to prevent this, but I figure it's better to address this at the module level to prevent it causing performance issues in other templates in the future :)

See also:
https://github.com/kiln/template-3d-point-map/issues/97#issuecomment-662986017

hughsk committed:
Improve performance of repeated calls to popup.hide()

Previously, _resizeConstrainer() was being called every time
popup.hide() was being called. This introduces a simple check
that prevents popup.hide() from running any heavy code if it's
already hidden.

This improves the performance of the popup module in the
point map template, which currently calls popup.hide() once
per frame. It's possible to make changes in the template to
prevent this, but I figure it's better to address this at the
module level to prevent it causing performance issues in other
templates in the future :)

See also:
https://github.com/kiln/template-3d-point-map/issues/97#issuecomment-662986017

daanlouter commented:
Nice change. Will have a look at this later.

I'm not 100% sure about this, but one thing to look out for is that currently .hide() also sets the new bounds of the constrainer. If `.hide()` is called in a generic `update()` loop then previously it made sure the constrainer would resize accordingly. It's possible that after this change the constrainer won't do this anymore. This means that when reducing window size, the constrainer can still take up more space than the window and add scrollbars.

daanlouter commented:
Had a quick play with this, and seems to work well, but the problem I was expecting did occur.

Basically if your code in update() has a snippet similar to this:
```js
if (state.sticky_popup) {
    popup.draw(bla_bla)
}
else {
    popup.hide()
}
```

The constrainer will start overlapping when reducing the window size.

Maybe we should just set the constrainer to `display: none` on hide()?

hughsk committed:
Hide popup constrainer when popup is hidden

This prevents the constrainer from extending beyond its bounds
when resizing the window when the popup is hidden.

hughsk commented:
Tested with sticky and non-sticky popups on the point map and `display: none` seems to do the trick! No issues arising either -- though may want to check again in another template :)

daanlouter commented:
Seems to work for me too 👍  Do you want to update version numbers?

hughsk committed:
2.1.3

hughsk commented:
Done! :)

daanlouter commented:
🥳
